toolchain/check/generic.cpp

```diff
@@ -147,21 +150,38 @@ class RebuildGenericConstantInEvalBlockCallbacks : public SubstInstCallbacks {
     auto dependence = orig_symbolic_const.dependence;
 
     // We might already have an instruction in the eval block if a transitive
-    // operand of this instruction has the same constant value.
-    auto result = constants_in_generic_.Insert(const_inst_id, [&] {
-      // TODO: Add a function on `Context` to add the instruction without
-      // inserting it into the dependent instructions list or computing a
-      // constant value for it.
-      // TODO: Is the location we pick here always appropriate for the new
-      // instruction?
-      auto inst_id = context_.sem_ir().insts().AddInNoBlock(
-          SemIR::LocIdAndInst::UncheckedLoc(loc_id_, new_inst));
-      auto const_id = AddGenericConstantInstToEvalBlock(
-          context_, generic_id_, region_, const_inst_id, inst_id, dependence);
-      context_.constant_values().Set(inst_id, const_id);
-      return inst_id;
-    });
-    return result.value();
+    // operand of this instruction has the same constant value, or from a
+    // previous declaration.
+    if (auto result = constants_in_generic_.Lookup(const_inst_id)) {
+      return result.value();
+    }
```

Comment:
I think it's worth moving this lookup into the `if (inside_redeclaration_)` below, so that we can continue to perform only a single hash table lookup in the other case.

Reply:
This is easier with it being FATAL, since we don't have to worry about inserting and then ... what. We can insert and then FATAL while trying to build the value.